openedx/core/djangoapps/course_views/course_views.py

```diff
@@ -0,0 +1,178 @@
+    title = None    # The title of the view, which should be internationalized
+    priority = None    # The relative priority of this view that affects the ordering (lower numbers shown first)
+    view_name = None    # The name of the Django view to show this view
+    tab_id = None    # The id to be used to show a tab for this view
+    is_movable = True    # True if this course view can be moved
+    is_dynamic = False    # True if this course view is dynamically added to the list of tabs
+    is_default = True    # True if this course view is a default for the course (when enabled)
+    is_hideable = False    # True if this course view's visibility can be toggled by the author
+
+    @classmethod
+    def is_enabled(cls, course, django_settings, user=None):  # pylint: disable=unused-argument
+        """Returns true if this course view is enabled in the course.
+
+        Args:
+            course (CourseDescriptor): the course using the feature
+            settings (dict): a dict of configuration settings
+
+            user (User): the user interacting with the course
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def validate(cls, tab_dict, raise_error=True):  # pylint: disable=unused-argument
+        """
+        Validates the given dict-type `tab_dict` object to ensure it contains the expected keys.
+        This method should be overridden by subclasses that require certain keys to be persisted in the tab.
+        """
+        return True
+
+    @classmethod
+    def create_tab(cls, tab_dict):
+        """
+        Returns the tab that will be shown to represent an instance of a view.
+        """
+        return CourseViewTab(cls, tab_dict=tab_dict)
+
+
+class CourseViewTypeManager(PluginManager):
+    """
+    Manager for all of the course view types that have been made available.
+
+    All course view types should implement `CourseViewType`.
+    """
+    NAMESPACE = COURSE_VIEW_TYPE_NAMESPACE
+
+    @classmethod
+    def get_course_view_types(cls):
+        """
+        Returns the list of available course view types in their canonical order.
+        """
+        def compare_course_view_types(first_type, second_type):
+            """Compares two course view types, for use in sorting."""
+            first_priority = first_type.priority
+            second_priority = second_type.priority
+            if not first_priority == second_priority:
+                if not first_priority:
+                    return 1
+                elif not second_priority:
+                    return -1
+                else:
+                    return first_priority - second_priority
+            first_name = first_type.name
+            second_name = second_type.name
+            if first_name < second_name:
+                return -1
+            elif first_name == second_name:
+                return 0
+            else:
+                return 1
```

Comment:
Do we have any tests of this ordering code? It seems like there are a lot of edge cases we'd like to make sure are behaving the way we expect.